_posts/2021-08-16-connector-table-sql-api-part2.md

```diff
@@ -0,0 +1,434 @@
+        alice@acme.org
+
+```
+
+The rows “Test Subject” should now have appeared as a row in your output. Our source is working!
+
+However, we’re still hard-coding the schema produced by the source, and e.g. defining the table with a different schema will produce errors. We want to be able to define which fields of an email interest us, however, and then produce the data accordingly. For this, we’ll use the list of column names we held onto earlier, and then simply look at it when we collect the emails. For brevity, we’ll only include a few of the possible fields here:
+
+```java
+private void collectMessages(SourceFunction.SourceContext<RowData> ctx, Message[] messages) {
+        for (Message message : messages) {
+            try {
+                collectMessage(ctx, message);
+            } catch (MessagingException ignored) {}
+        }
+    }
+
+    private void collectMessage(SourceFunction.SourceContext<RowData> ctx, Message message)
+        throws MessagingException {
+        var row = new GenericRowData(columnNames.size());
+
+        for (int i = 0; i < columnNames.size(); i++) {
+            switch (columnNames.get(i)) {
+                case "SUBJECT":
+                    row.setField(i, StringData.fromString(message.getSubject()));
+                    break;
+                case "SENT":
+                    row.setField(i, TimestampData.fromInstant(message.getSentDate().toInstant()));
+                    break;
+                case "RECEIVED":
+                    row.setField(i, TimestampData.fromInstant(message.getReceivedDate().toInstant()));
+                    break;
+                // ...
+            }
+        }
+
+        ctx.collect(row);
+    }
+```
+
+You should now have a working source that we can select any of the columns from which we support. We can try it out once again, but this time specifying all the columns we support above:
+
+```sql
+CREATE TABLE T (
+    subject STRING,
+    sent TIMESTAMP(3),
+    received TIMESTAMP(3)
+) WITH (
+    'connector' = 'imap', 
+    'host' = 'greenmail',
+    'port' = '3143', 
+    'user' = 'alice', 
+    'password' = 'alice'
+);
+
+SELECT * FROM T;
+```
+
+Use the command from earlier to send emails to the greenmail server and you should see them appear. You can also try selecting only some of the columns, or write more complex queries. Note, however, that there are quite a few more things we haven’t covered here, such as advancing watermarks.
+
+
+# Test the connector with a real mail server on the Ververica Platform 
+
+If you want to test the connector with a real mail server, you can import it into [Ververica Platform Community Edition](https://www.ververica.com/getting-started). 
+
+Since our example connector in this blog post is still a bit limited, we’ll actually use github.com/Airblader/flink-connector-imap instead this time. We’ll also assume you already have Ververica Platform up and running (see the link above).
```

Comment:
Rename this to flink-connector-email and convert it into a link.

As of tomorrow the link will also be `github.com/TNG/flink-connector-email` instead, but in theory the old URL should keep working.